### ISSUE TITLE
[Sub-object] Prototype sub-object selection

### ISSUE THREAD

**Issue by VWoeltjen** [nasa/openmct#1126]
See #1101 for design. Implement sub-object selection at least to prototype quality; in particular, want to look at:
- What APIs need to be exposed? How should these be exposed?
- What is minimal functionality necessary for this to be validated with users? (See notes in #1101, this has been discussed/scoped already)

**Comment by VWoeltjen** [comment]
Notes on scope & initial approach from discussion on 7/27/2016:

> * Approach
>   * Bring in new selection API
>   * Use Selection from Inspector
>   * Make tree Select and Navigate
>   * Support Select from Plot

**Comment by VWoeltjen** [comment]
Think that Selection API from tutorial prototypes should be evolved here as well; we want the API to be flexible enough to support multi-select and selection context, whereas that API is rather minimal (single selection, no context)

**Comment by VWoeltjen** [comment]
Reference list of selection APIs, from web searching:

[SketchUp](https://www.sketchup.com/intl/en/developer/docs/ourdoc/selection)
[Unity](https://docs.unity3d.com/ScriptReference/Selection.html)
[NetBeans](https://platform.netbeans.org/tutorials/nbm-selection-1.html)
[LibGDX](https://libgdx.badlogicgames.com/nightlies/docs/api/com/badlogic/gdx/scenes/scene2d/utils/Selection.html)
[Eclipse](https://eclipse.org/articles/Article-WorkbenchSelections/article.html)

**Comment by VWoeltjen** [comment]
| Intuitive | Callbacky | Multiple | Context  | Decoupled
---------|-----------|-----------|----------|----------|-----------
[SketchUp](https://www.sketchup.com/intl/en/developer/docs/ourdoc/selection) | :white_check_mark:       | :white_check_mark:       | :white_check_mark:      | :x:     | :x:
[Unity](https://docs.unity3d.com/ScriptReference/Selection.html)    | :white_check_mark:       | :white_check_mark:       | :white_check_mark:      | Indirect | :x:
[NetBeans](https://platform.netbeans.org/tutorials/nbm-selection-1.html) | :x:      | :white_check_mark:       | :grey_question:    | Indirect | :white_check_mark:
[LibGDX](https://libgdx.badlogicgames.com/nightlies/docs/api/com/badlogic/gdx/scenes/scene2d/utils/Selection.html)   | :white_check_mark:       | :white_check_mark:       | :white_check_mark:      | :x:     | :white_check_mark:
[Eclipse](https://eclipse.org/articles/Article-WorkbenchSelections/article.html)  | :x:      | :white_check_mark:       | :white_check_mark:      | :white_check_mark:      | :white_check_mark:

* "Intuitive" is subjective; NetBeans and Eclipse both have selection APIs
  which require lengthy documentation and refer to other APIs, so using
  these APIs seems like it would take more than just a quick skim through
  documentation. Other APIs are much more straightforward.
* "Callbacky" is just a fancy word for "fires change events." All the APIs
  here do so, and we were planning to do so, so that decision seems 
  validated.
* "Multiple" refers to multi-select. We want an API that supports that 
  while introducing minimal complexity to the simple case. It is unclear
  to me from NetBeans documentation whether or not this is supported.
  (Selections are modelled as "Lookups", which look powerful/effective 
  but not very intuitive.)
* "Context" refers to the notion we have of a selection having some path.
  For instance, the "Remove" action would actually depend on the parent 
  of the selection (so it needs to know that path.) In whiteboarding 
  sessions, we've discussed handling this as an array and/or as a series
  of arguments to some method call. Only Eclipse includes this in the 
  selection API (if I understand `ITreeSelection.getPaths()` correctly)
  and is then only one of several selection models. This is a bit of a 
  red flag; should consider options to decouple selection from context. 
  In some cases, there is an object hierarchy that can be accessed from 
  the objects that get selected (which would be more similar to our 
  `getCapability('context')` approach.)
* "Decoupled" is used to designate APIs which make minimal assumptions
  about what can be selected. APIs are considered decoupled if selections
  may be of arbitrary types, or if the type is non-arbitrary but can 
  trivally wrap an arbitrary type.
  
SketchUp/LibGDX/Unity look like the models we want to follow; NetBeans/Eclipse have thoroughly-engineered selection models that look like they handle a broad range of cases well, but also impose a significant learning curve. We want to be simpler than an RCP.

The one thing we can't easily borrow from these is a solution for selection context. If this responsibility were moved to selected objects themselves, then this ceases to be a problem, but this also imposes restrictions on what can be selected.

**Comment by larkin** [comment]
I like looking at the other APIs for insights here.  At the same time, I would ask: "What portion of 'selection' is part of the public API versus the private API?"  

Publicly,  I think developers need to be able to:

1. specify that nodes in their views are selectable
2. register views to show in the inspector in response to certain selections?
3. register views to show in the toolbar in response to certain selections?
4. register actions available for a given selection (which populate context menus)

Does this seem reasonable or is there a larger set of things we'd like them to do with selection?  And is it okay to place a limit on the flexibility of our selection API?

On context, I'm thinking that ties into the first point above.  We implement a method like `MCT.makeSelectable(domElement, object)` which adds selection functionality to a given `domElement` (css classes, gesture handlers, etc).  To handle context, on selection, MCT climbs the tree to build the context path for the object.

For 2-4 if we stick with our initial concept of taking the path array, reversing it, and `apply`ing it to the `canView` function for a view or the `canAct` function for an action, we then allow view and action definitions to utilize context by accepting multiple arguments.  The default behavior is "I don't care about context" but it's easy to opt in.

**Comment by VWoeltjen** [comment]
Very good point with the developer use cases - I was thinking naively of "plot uses selection" and going from there, but it is much nicer to say "plot says this legend entry is selectable" and let the platform handle it.

With the approach you describe, I think we can hide the selection API completely, particularly if toolbar and inspector are both views of the primary selection. The multi-select case complicates things somewhat; batching things works for at least some actions, but not for views. Treating views as views of a "composite domain object" could plausibly work, but things would get funny around indeterminate state. (Multi-select is out of scope for this issue, but don't want to move forward on a solution that will be incompatible.)

**Comment by larkin** [comment]
Hrm, the multi-select dillema again.  Some experiments for methods of handling that:

```javascript
// A) Keep current naming for singular selection, use new methods for multiple selection.
// checking for applicability (returns true/false)
canView = function (object, parent, grandparent) {};
canViewMultiple = function ([objectA, objectB], [parentA, parentB], [grandparentA, grandparentB]...) {};
// instantiating a view (returns view instance)
view = function (object, parent, grandparent...) {};
viewMultiple = function ([objectA, objectB], [parentA, parentB], [grandparentA, grandparentB]...) {};

// B) Overload existing functions and make it harder to register single views.
// check for applicability, return true/false
canView = function ([objectA, objectB], [parentA, parentB], [grandparentA, grandparentB]...) {};
// instantiating a view (returns view instance)
view = function ([objectA, objectB], [parentA, parentB], [grandparentA, grandparentB]...) {};

// C) perform view checking against singular selections multiple times, but define custom factory
// for rendering a view of multiple objects.  use inspection to determine if a view is valid for 
// multi selection (e.g. if all objects can be viewed, and view defines viewMultiple)
// determines if view can apply to an object.
canView = function (object, parent, grandparent...) {}; 
// view factories return view instances.
view = function (object, parent, grandparent...) {};
viewMultiple = function ([objectA, objectB], [parentA, parentB], [grandparentA, grandparentB]...) {};
```

The same principles could work with Actions, but in the case of actions you don't necessarily need to implement different behavior to execute an action on a multiple selection, we could just batch them all in a transaction to ensure consistent pass/fail behavior.  Though perhaps we could add functionality down the road to construct labels on demand, e.g. "remove" for a single item or "remove 2 items" with 2 items selected.

**Comment by larkin** [comment]
With regards to hiding the API completely: maybe we just put it in the background with a "provisional" flag, and expose functionality so that people generally don't need to use it.  If we can find a way where no one needs to use it, then we remove it.

The uses cases I'm looking at to flesh out selection are visible in two places: the todo tutorial and the fixed position view.  In these cases, The toolbar always has a "add" button, but it also gains a "delete" button when an object is selected.

Some questions for further thought-- do we layer selection on top of regions?  i.e.
1. views or actions can be registered to the selection and will display in the inspector/toolbar/etc.
2. views registered to the toolbar or inspector region override default selection behavior.

**Comment by VWoeltjen** [comment]
Work in progress on branch `subobject-selection-1126`

**Comment by VWoeltjen** [comment]
One issue with the `foo(item, parent, grandparent...)` approach is that it doesn't give any information about context, such as the element that initiated the action. That limitation was problematic in #1164. It also isn't extensible (barring weird games with global state) to allow more information to be passed in the future.

Due to this, I'm inclined to move to a `foo(item, context)` approach, where Context has the following properties:

* `item`: Redundant to the item passed in
* `parent`: The parent Context (or undefined)
* `element`: The HTML element (if any)
* `event`:  The event which triggered this action (if any)

We're in general agreement that context objects suck to deal with, but this at least reduces to `foo(item)` for common cases. Useful methods like `path()` (get an array of all the items)

@larkin, any holes you can poke here?

**Comment by larkin** [comment]
@VWoeltjen It seems like using the context object is the common case, not the exception; thus we might want to just do `foo(context)`.  YMMV.

There's the question of "what is the selection api," and the more pressing questions of "what is our API for managing context menu actions" and "what is our API for managing view changes based on selection."

We'd discussed that selection is a precursor to a context menu.  Maybe we should prototype an overhaul of the context menu based on the selection API, so that we can start seeing this in usage and work out some of the details?

**Comment by larkin** [comment]
High priority for Roddenberry.

**Comment by VWoeltjen** [comment]
Random relevant class diagram found lying around

![image](https://user-images.githubusercontent.com/1685764/30340217-9751cab6-97a6-11e7-9272-8690f3ed80d5.png)

Looks like a good approach but I don't think "item" belongs in a generic Gesture interface. (The Gesture API is also a [separate problem](https://github.com/nasa/openmct/pull/1653) in any event.)

**Comment by VWoeltjen** [comment]
Notes from discussing 10/17/2017, example of using a context object (e.g. from a layout):

```
openmct.selection.add({
    item: myDomainObject,
    parent: layoutContext
    element: myLayoutFrame
});
```

**Comment by VWoeltjen** [comment]
Implemented and/or made obsolete by #1811 et al
